### PR TITLE
Vault-level RPCs

### PR DESCRIPTION
**Pull request by nbaksalyar**
This change is to have full information about a requester (if it's an app) at the destination elders.

### CODE REVIEW

File: src/destination_elder.rs [hunk 1]
```diff
@@ -319,10 +332,19 @@ impl DestinationElder {
...
     ) -> Option<Action> {
+        let requester_pk = match requester {
+            PublicId::Client(ref client) => client.public_key(),
+            PublicId::App(ref app) => app.public_key(),
+            PublicId::Node(..) => {
+                // TODO: drop the connection, possible malice
+                warn!("Unexpected PublicId::Node requester: {:?}", requester);
+                return None;
+            }
+        };
```
Comment (madadam) [inline src/destination_elder.rs]: Might be worth extracting this into a function. The same snippet is used in https://github.com/maidsafe/safe_vault/blob/315c3a88c49447ac98e51326d53fba24259d1dec/src/source_elder.rs#L378-L385, https://github.com/maidsafe/safe_vault/blob/315c3a88c49447ac98e51326d53fba24259d1dec/src/source_elder.rs#L472-L483 and possibly elsewhere.

File: src/destination_elder.rs [hunk 2]
```diff
@@ -369,18 +391,27 @@ impl DestinationElder {
...
```
Comment (madadam) [inline src/destination_elder.rs]: Ditto.

File: src/destination_elder.rs [hunk 3]
```diff
@@ -628,14 +670,23 @@ impl DestinationElder {
...
+        let client = self.client_id(&message_id)?;
+        let client_pk = match client {
+            PublicId::Client(ref client) => client.public_key(),
+            PublicId::App(ref app) => app.public_key(),
+            PublicId::Node(..) => {
+                // TODO: drop the connection, possible malice
+                warn!("Unexpected PublicId::Node requester: {:?}", client);
+                return None;
+            }
+        };
```
Comment (madadam) [inline src/destination_elder.rs]: Ditto

File: src/destination_elder.rs [hunk 4]
```diff
@@ -652,16 +703,26 @@ impl DestinationElder {
...
-        if &src == address.name() {
+        let requester_pk = match requester {
+            PublicId::Client(ref client) => client.public_key(),
+            PublicId::App(ref app) => app.public_key(),
+            PublicId::Node(..) => {
+                // TODO: drop the connection, possible malice
+                warn!("Unexpected PublicId::Node requester: {:?}", requester);
+                return None;
+            }
+        };
```
Comment (madadam) [inline src/destination_elder.rs]: Ditto

File: src/destination_elder.rs [hunk 5]
```diff
@@ -698,15 +762,24 @@ impl DestinationElder {
...
+        let client = self.client_id(&message_id)?;
+        let client_pk = match client {
+            PublicId::Client(ref client) => client.public_key(),
+            PublicId::App(ref app) => app.public_key(),
+            PublicId::Node(..) => {
+                // TODO: drop the connection, possible malice
+                warn!("Unexpected PublicId::Node requester: {:?}", client);
+                return None;
+            }
+        };
```
Comment (madadam) [inline src/destination_elder.rs]: Ditto

File: src/destination_elder.rs (hunk 1 above)
Comment (nbaksalyar) [reply src/destination_elder.rs]: Yeah, I thought that it probably should be a part of `safe-nd`, but it's not very clear how to handle `PublicId::App` - there are no universal rules.

So maybe it should be 2 separate functions, like `PublicId::owner_public_key` and `PublicId::own_public_key`? @Fraser999 what's your opinion on this?

Comment (madadam) [reply src/destination_elder.rs]: :+1: for the two functions.

Comment (madadam) [reply src/destination_elder.rs]: Don't think it needs to go to safe-nd necessarily. Unless the client libs have a use for it too?

Comment (Fraser999) [reply src/destination_elder.rs]: Yeah, I'd keep it out of safe-nd if possible.  But +1 from me on the two helpers.  The main downside I see is that the `Node` variant should return `None` for our use cases, but the name of the helpers doesn't really suggest that.

File: src/destination_elder.rs [hunk 6]
```diff
@@ -118,9 +119,19 @@ impl DestinationElder {
...
+        match message {
+            VaultMessage::ClientRequest {
+                request,
+                requester,
+                message_id,
+            } => self.handle_request(requester, request, message_id),
+        }
+    }
+
     pub fn handle_request(
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/destination_elder.rs]: It does seem like we should choose between handling safe_nd-level messages or vault-level ones in the persona APIs.

File: src/destination_elder.rs [hunk 7]
```diff
@@ -787,6 +852,6 @@ impl DestinationElder {
 
 impl Display for DestinationElder {
     fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
-        write!(formatter, "{}", self.id.name())
+        write!(formatter, "{}", self.id)
```
Comment (Fraser999) [inline src/destination_elder.rs]: Did you mean to revert this recent change?

**Pull request merged**